batcher/aligned-sdk/src/communication/messaging.rs

```diff
@@ -55,178 +55,198 @@ pub async fn send_messages(
+                    + &merkle_root
+                    + ", failed with error: "
+                    + &error,
+            ))
         }
         Ok(ResponseMessage::ProtocolVersion(_)) => {
-            return Err(SubmitError::UnexpectedBatcherResponse(
+            error!("Batcher responded with protocol version instead of batch inclusion data");
+            Err(SubmitError::UnexpectedBatcherResponse(
                 "Batcher responded with protocol version instead of batch inclusion data"
                     .to_string(),
-            ));
+            ))
         }
         Ok(ResponseMessage::BatchReset) => {
-            return Err(SubmitError::ProofQueueFlushed);
+            error!("Batcher responded with batch reset");
+            Err(SubmitError::ProofQueueFlushed)
         }
         Ok(ResponseMessage::Error(e)) => {
             error!("Batcher responded with error: {}", e);
-        }
-        Ok(ResponseMessage::CreateNewTaskError(merkle_root)) => {
-            return Err(SubmitError::BatchSubmissionFailed(
-                "Could not create task with merkle root ".to_owned() + &merkle_root,
-            ));
-        }
-        Ok(ResponseMessage::InvalidProof(reason)) => {
-            return Err(SubmitError::InvalidProof(reason));
+            Err(SubmitError::GenericError(e))
         }
         Err(e) => {
-            return Err(SubmitError::SerializationError(e));
+            error!("Error while deserializing batch inclusion data: {}", e);
+            Err(SubmitError::SerializationError(e))
+        }
+    }
+}
+
+// Used to match the message received from the batcher,
+// with the NoncedVerificationData you sent
+// This is used to verify the proof you sent was indeed included in the batch
+fn match_batcher_response_with_stored_verification_data(
+    batch_inclusion_data: &BatchInclusionData,
+    sent_verification_data: &mut Vec<NoncedVerificationData>,
+) -> Result<VerificationDataCommitment, SubmitError> {
+    debug!("Matching verification data with batcher response ...");
+    let mut index = None;
+    for (i, sent_nonced_verification_data) in sent_verification_data.iter_mut().enumerate() {
+        if sent_nonced_verification_data.nonce == batch_inclusion_data.user_nonce {
+            debug!("local nonced verification data matched with batcher response");
+            index = Some(i);
+            break;
         }
     }
 
-    Ok(())
+    if let Some(i) = index {
+        let verification_data = sent_verification_data.swap_remove(i); //TODO maybe only remove?
+        return Ok(verification_data.verification_data.clone().into());
+    }
+
+    Err(SubmitError::InvalidProofInclusionData)
+}
+
+// Returns the biggest nonce from the sent verification data
+// Used to know which is the last proof sent to the Batcher,
+// to know when to stop reading the WS for responses
+fn get_biggest_nonce(sent_verification_data: &[NoncedVerificationData]) -> U256 {
```

Comment:
We should be able to get the biggest nonce from the latest element, right?

Reply:
yes but not necesarrily. replacement messages can come after the biggest one, and they will be responsded in nonce order